File: packages/ac-schema/src/mcp.ts
import { z } from 'zod';

// MCP Server Definition (read from host's config.toml, secrets redacted)
export const MCPServerSchema = z.object({
  name: z.string(),
  display_name: z.string().optional(),
  description: z.string().optional(),
  transport: z.enum(['stdio', 'http']).default('stdio'),
  // For stdio MCPs
  command: z.string().optional(),
  args: z.array(z.string()).optional(),
  // For http MCPs
  url: z.string().optional(),
  // Env vars - secret values are redacted to "***REDACTED***"
  env: z.record(z.string()).optional(),
  // Flags
  has_secrets: z.boolean().default(false),
  poolable: z.boolean().default(false),
});
export type MCPServer = z.infer<typeof MCPServerSchema>;

// MCP enablement for a session or project
export const MCPEnablementSchema = z.object({
  enabled: z.boolean(),
  scope: z.enum(['session', 'project', 'global']).default('session'),
});
export type MCPEnablement = z.infer<typeof MCPEnablementSchema>;

// MCP config for a session (returned by GET /v1/sessions/:id/mcp)
export const SessionMCPConfigSchema = z.object({
  session_id: z.string().uuid(),
  servers: z.array(MCPServerSchema),
  enablement: z.record(MCPEnablementSchema), // keyed by mcp name
  restart_required: z.boolean().default(false),
});
export type SessionMCPConfig = z.infer<typeof SessionMCPConfigSchema>;

// Request to update MCP config for a session
export const UpdateMCPConfigRequestSchema = z.object({
  enablement: z.record(z.object({
    enabled: z.boolean(),
    scope: z.enum(['session', 'project', 'global']).optional(),
  })),
});
export type UpdateMCPConfigRequest = z.infer<typeof UpdateMCPConfigRequestSchema>;

// MCP Pool configuration
export const MCPPoolConfigSchema = z.object({
  enabled: z.boolean().default(false),
  pool_all: z.boolean().default(false),
  exclude_mcps: z.array(z.string()).optional(),
});
export type MCPPoolConfig = z.infer<typeof MCPPoolConfigSchema>;

// Messages for MCP operations
export const MCPListRequestMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.list_servers'),
  ts: z.string().datetime({ offset: true }),
  payload: z.object({
    cmd_id: z.string(),
    host_id: z.string().uuid(),
  }),
});

export const MCPGetConfigMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.get_config'),
  ts: z.string().datetime({ offset: true }),
  payload: z.object({
    cmd_id: z.string(),
    session_id: z.string().uuid(),
  }),
});

export const MCPUpdateConfigMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.update_config'),
  ts: z.string().datetime({ offset: true }),
  payload: z.object({
    cmd_id: z.string(),
    session_id: z.string().uuid(),
    enablement: z.record(z.object({
      enabled: z.boolean(),
      scope: z.enum(['session', 'project', 'global']).optional(),
    })),
  }),
});

export const MCPServersResponseMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.servers'),
  ts: z.string().datetime({ offset: true }),
  seq: z.number().int().positive(),
  payload: z.object({
    cmd_id: z.string(),
    servers: z.array(MCPServerSchema),
    pool_config: MCPPoolConfigSchema.optional(),
  }),
});

export const MCPConfigResponseMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.config'),
  ts: z.string().datetime({ offset: true }),
  seq: z.number().int().positive(),
  payload: SessionMCPConfigSchema.extend({
    cmd_id: z.string(),
  }),
});

export const MCPProjectConfigResponseMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.project_config'),
  ts: z.string().datetime({ offset: true }),
  seq: z.number().int().positive(),
  payload: z.object({
    cmd_id: z.string(),
    enablement: z.record(MCPEnablementSchema),
  }),
});

export const MCPUpdateResultMessageSchema = z.object({
  v: z.literal(1),
  type: z.literal('mcp.update_result'),
  ts: z.string().datetime({ offset: true }),
  seq: z.number().int().positive(),
  payload: z.object({
    cmd_id: z.string(),
    success: z.boolean(),
    restart_required: z.boolean(),
    error: z.string().optional(),
  }),
});

File: agents/agentd/internal/protocol/mcp.go
package protocol

type MCPEnablement struct {
	Enabled bool   `json:"enabled"`
	Scope   string `json:"scope,omitempty"`
}

type MCPServer struct {
	Name        string            `json:"name"`
	DisplayName string            `json:"display_name,omitempty"`
	Description string            `json:"description,omitempty"`
	Transport   string            `json:"transport"`
	Command     string            `json:"command,omitempty"`
	Args        []string          `json:"args,omitempty"`
	URL         string            `json:"url,omitempty"`
	Env         map[string]string `json:"env,omitempty"`
	HasSecrets  bool              `json:"has_secrets"`
	Poolable    bool              `json:"poolable"`
}

type MCPPoolConfig struct {
	Enabled     bool     `json:"enabled"`
	PoolAll     bool     `json:"pool_all"`
	ExcludeMCPs []string `json:"exclude_mcps,omitempty"`
}

type MCPListServersPayload struct {
	CmdID  string `json:"cmd_id"`
	HostID string `json:"host_id"`
}

type MCPGetConfigPayload struct {
	CmdID     string `json:"cmd_id"`
	SessionID string `json:"session_id"`
}

type MCPUpdateConfigPayload struct {
	CmdID      string                   `json:"cmd_id"`
	SessionID  string                   `json:"session_id"`
	Enablement map[string]MCPEnablement `json:"enablement"`
}

type MCPGetProjectConfigPayload struct {
	CmdID    string `json:"cmd_id"`
	RepoRoot string `json:"repo_root"`
}

type MCPUpdateProjectConfigPayload struct {
	CmdID      string                   `json:"cmd_id"`
	RepoRoot   string                   `json:"repo_root"`
	Enablement map[string]MCPEnablement `json:"enablement"`
}

type MCPServersPayload struct {
	CmdID      string        `json:"cmd_id"`
	Servers    []MCPServer   `json:"servers"`
	PoolConfig MCPPoolConfig `json:"pool_config,omitempty"`
}

type MCPConfigPayload struct {
	CmdID           string                   `json:"cmd_id"`
	SessionID       string                   `json:"session_id"`
	Servers         []MCPServer              `json:"servers"`
	Enablement      map[string]MCPEnablement `json:"enablement"`
	RestartRequired bool                     `json:"restart_required"`
}

type MCPProjectConfigPayload struct {
	CmdID      string                   `json:"cmd_id"`
	Enablement map[string]MCPEnablement `json:"enablement"`
}

type MCPUpdateResultPayload struct {
	CmdID           string `json:"cmd_id"`
	Success         bool   `json:"success"`
	RestartRequired bool   `json:"restart_required"`
	Error           string `json:"error,omitempty"`
}
